File: frontend/src/state/qbank.ts
// QBank Store —— 题库与练习状态管理
// ------------------------------------------------------------
// 对接后端 Wails 绑定：
// - QBankExtract(uri, title) —— 从文档抽题生成题集
// - QBankSave(set) —— 保存题集
// - QBankStartAttempt(setID) —— 开始一次答题
// - QBankAnswer(attemptID, qID, ans) —— 提交单题答案
// - QBankSubmit(attemptID) —— 提交整卷评分
// - QBankAnalyze(setID, qID) —— 单题 AI 解析
// - QBankMastery() —— 知识点掌握度映射

import { create } from "zustand";
import { callWails } from "@/lib/wails";

/** 题目类型：单选 / 多选 / 填空 / 简答 */
export type QuestionType = "single" | "multi" | "fill" | "essay";

/** 题目结构 —— 与后端 qbank.Question 对齐 */
export interface Question {
  id: string;
  stem: string;
  options?: string[];
  answer: string;
  type: string;
  points?: string[];
  knowledge?: string[];
}

/** 题集结构 —— 与后端 qbank.Set 对齐 */
export interface QBSet {
  id: string;
  title: string;
  questions: Question[];
  created_at?: unknown;
}

/** 答题记录 —— 与后端 qbank.Attempt 对齐 */
export interface Attempt {
  id: string;
  set_id: string;
  answers: Record<string, string>;
  score: number;
  total: number;
  started_at?: unknown;
  finished_at?: unknown;
}

interface QBankState {
  /** 已保存的题集列表 */
  sets: QBSet[];
  /** 当前选中的题集 ID */
  activeSetId: string | null;
  /** 当前题集对象（来自 sets 缓存或最近 Extract 的结果） */
  activeSet: QBSet | null;
  /** 当前答题 attempt */
  attempt: Attempt | null;
  /** 当前题目索引（0-based） */
  currentIndex: number;
  /** 用户作答草稿（qID -> 答案字符串） */
  draftAnswers: Record<string, string>;
  /** 知识点掌握度映射 */
  mastery: Record<string, number>;
  /** 单题 AI 解析缓存（qID -> 解析文本） */
  analysisMap: Record<string, string>;
  /** 加载状态：通用 */
  loading: boolean;
  /** 加载状态：抽题 */
  extracting: boolean;
  /** 加载状态：提交评分 */
  submitting: boolean;
  /** 加载状态：AI 解析（按 qID 索引） */
  analyzingQid: string | null;
  /** 错误信息 */
  error: string | null;

  // —— Actions ——
  extract: (uri: string, title: string) => Promise<QBSet | null>;
  save: (set: QBSet) => Promise<string | null>;
  selectSet: (id: string) => void;
  startAttempt: () => Promise<void>;
  answer: (qID: string, ans: string) => Promise<void>;
  setDraft: (qID: string, ans: string) => void;
  submit: () => Promise<void>;
  analyze: (qID: string) => Promise<void>;
  loadMastery: () => Promise<void>;
  next: () => void;
  prev: () => void;
  jumpTo: (idx: number) => void;
  reset: () => void;
}

export const useQBankStore = create<QBankState>((set, get) => ({
  sets: [],
  activeSetId: null,
  activeSet: null,
  attempt: null,
  currentIndex: 0,
  draftAnswers: {},
  mastery: {},
  analysisMap: {},
  loading: false,
  extracting: false,
  submitting: false,
  analyzingQid: null,
  error: null,

  // —— 从 URI 抽题 ——
  extract: async (uri, title) => {
    set({ extracting: true, error: null });
    try {
      const result = await callWails<QBSet>("QBankExtract", uri, title);
      if (!result) {
        set({ error: "[后端未连接] 抽题不可用" });
        return null;
      }
      // 抽题成功后立即设为当前题集（用户可选择保存）
      set({
        activeSet: result,
        activeSetId: result.id,
        attempt: null,
        currentIndex: 0,
        draftAnswers: {},
        analysisMap: {},
      });
      return result;
    } catch (err) {
      set({ error: err instanceof Error ? err.message : String(err) });
      return null;
    } finally {
      set({ extracting: false });
    }
  },

  // —— 保存题集 ——
  save: async (qbSet) => {
    set({ loading: true, error: null });
    try {
      const uri = await callWails<string>("QBankSave", qbSet);
      if (uri) {
        // 保存成功后追加到本地列表（去重）
        set((s) => ({
          sets: s.sets.some((x) => x.id === qbSet.id)
            ? s.sets.map((x) => (x.id === qbSet.id ? qbSet : x))
            : [...s.sets, qbSet],
        }));
      }
      return uri;
    } catch (err) {
      set({ error: err instanceof Error ? err.message : String(err) });
      return null;
    } finally {
      set({ loading: false });
    }
  },

  // —— 选中题集 ——
  selectSet: (id) => {
    const set_ = get().sets.find((s) => s.id === id) ?? null;
    set({
          activeSetId: id,
          activeSet: set_,
          attempt: null,
          currentIndex: 0,
          draftAnswers: {},
          analysisMap: {},
        });
  },

  // —— 开始一次答题 ——
  startAttempt: async () => {
    const { activeSetId } = get();
    if (!activeSetId) {
      set({ error: "未选中题集" });
      return;
    }
    set({ loading: true, error: null });
    try {
      const att = await callWails<Attempt>("QBankStartAttempt", activeSetId);
      if (!att) {
        set({ error: "[后端未连接] 无法开始答题" });
        return;
      }
      set({
        attempt: att,
        currentIndex: 0,
        draftAnswers: { ...att.answers },
      });
    } catch (err) {
      set({ error: err instanceof Error ? err.message : String(err) });
    } finally {
      set({ loading: false });
    }
  },

  // —— 草稿作答（不立即提交，仅本地保存） ——
  setDraft: (qID, ans) => {
    set((s) => ({ draftAnswers: { ...s.draftAnswers, [qID]: ans } }));
  },

  // —— 提交单题答案到后端 ——
  answer: async (qID, ans) => {
    const { attempt } = get();
    if (!attempt) return;
    // 先更新本地草稿
    set((s) => ({ draftAnswers: { ...s.draftAnswers, [qID]: ans } }));
    // 调用后端持久化（失败不阻塞，最终以 submit 为准）
    await callWails<void>("QBankAnswer", attempt.id, qID, ans);
  },

  // —— 提交整卷 ——
  submit: async () => {
    const { attempt, draftAnswers } = get();
    if (!attempt) return;
    set({ submitting: true, error: null });
    try {
      // 先把所有草稿答案同步到后端
      for (const [qID, ans] of Object.entries(draftAnswers)) {
        await callWails<void>("QBankAnswer", attempt.id, qID, ans);
      }
      const finalAttempt = await callWails<Attempt>(
        "QBankSubmit",
        attempt.id
      );
      if (finalAttempt) {
        set({
          attempt: { ...finalAttempt, answers: { ...draftAnswers } },
        });
      } else {
        set({ error: "[后端未连接] 提交评分不可用" });
      }
    } catch (err) {
      set({ error: err instanceof Error ? err.message : String(err) });
    } finally {
      set({ submitting: false });
    }
  },

  // —— 单题 AI 解析 ——
  analyze: async (qID) => {
    const { activeSetId } = get();
    if (!activeSetId) return;
    set({ analyzingQid: qID, error: null });
    try {
      const text = await callWails<string>(
        "QBankAnalyze",
        activeSetId,
        qID
      );
      set((s) => ({
        analysisMap: {
          ...s.analysisMap,
          [qID]: text ?? "[后端未连接] 解析不可用",
        },
      }));
    } catch (err) {
      set((s) => ({
        analysisMap: {
          ...s.analysisMap,
          [qID]: "[错误] " + (err instanceof Error ? err.message : String(err)),
        },
      }));
    } finally {
      set({ analyzingQid: null });
    }
  },

  // —— 加载知识点掌握度 ——
  loadMastery: async () => {
    set({ loading: true, error: null });
    try {
      const m = await callWails<Record<string, number>>("QBankMastery");
      set({ mastery: m ?? {} });
    } catch (err) {
      set({ error: err instanceof Error ? err.message : String(err) });
    } finally {
      set({ loading: false });
    }
  },

  // —— 题目导航 ——
  next: () =>
    set((s) => ({
      currentIndex: Math.min(
        s.currentIndex + 1,
        Math.max(0, (s.activeSet?.questions.length ?? 1) - 1)
      ),
    })),
  prev: () => set((s) => ({ currentIndex: Math.max(0, s.currentIndex - 1) })),
  jumpTo: (idx) =>
    set((s) => ({
      currentIndex: Math.max(
        0,
        Math.min(idx, Math.max(0, (s.activeSet?.questions.length ?? 1) - 1))
      ),
    })),

  // —— 重置答题 ——
  reset: () =>
    set({
      attempt: null,
      currentIndex: 0,
      draftAnswers: {},
      analysisMap: {},
    }),
}));

File: internal/qbank/qbank.go
// Package qbank 题库：解析、练习、阅卷、历史、AI 解析、掌握度。
package qbank

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"

	"github.com/helixnow/deep-student-go/pkg/llm"
	"github.com/helixnow/deep-student-go/pkg/store"
	"github.com/helixnow/deep-student-go/pkg/vfs"
)

// Question 单题。
type Question struct {
	ID        string   `json:"id"`
	Stem      string   `json:"stem"`
	Options   []string `json:"options,omitempty"`
	Answer    string   `json:"answer"`
	Type      string   `json:"type"` // single | multi | fill | essay
	Points    []string `json:"points,omitempty"`
	Knowledge []string `json:"knowledge,omitempty"`
}

// Set 题目集。
type Set struct {
	ID        string     `json:"id"`
	Title     string     `json:"title"`
	Questions []Question `json:"questions"`
	CreatedAt time.Time  `json:"created_at"`
}

// Attempt 一次练习记录。
type Attempt struct {
	ID         string            `json:"id"`
	SetID      string            `json:"set_id"`
	Answers    map[string]string `json:"answers"`
	Score      int               `json:"score"`
	Total      int               `json:"total"`
	StartedAt  time.Time         `json:"started_at"`
	FinishedAt time.Time         `json:"finished_at"`
}

// Service 题库服务。
type Service struct {
	vfs      *vfs.FS
	store    *store.Store
	llm      *llm.Registry
	mu       sync.RWMutex
	sets     map[string]*Set
	attempts map[string]*Attempt
	mastery  map[string]int // knowledge -> score 0..100
}

// New 创建 Service。
func New(fs *vfs.FS, st *store.Store, l *llm.Registry) *Service {
	return &Service{
		vfs: fs, store: st, llm: l,
		sets: map[string]*Set{}, attempts: map[string]*Attempt{},
		mastery: map[string]int{},
	}
}

// Extract 从教材/试卷 URI 提取题目。
func (s *Service) Extract(ctx context.Context, sourceURI, title string) (*Set, error) {
	data, _, err := s.vfs.Get(sourceURI)
	if err != nil {
		return nil, err
	}
	p, ok := s.llm.Get("openai")
	if !ok {
		return nil, fmt.Errorf("qbank: no provider")
	}
	resp, err := p.Chat(ctx, llm.ChatRequest{
		Model: "gpt-4o-mini",
		Messages: []llm.Message{
			{Role: llm.RoleSystem, Content: "Extract all questions from the given text. Output JSON only."},
			{Role: llm.RoleUser, Content: "Text:\n" + truncate(string(data), 12000) + "\n\nReturn JSON: {\"questions\":[{\"id\":\"q1\",\"stem\":\"...\",\"options\":[\"A\",\"B\"],\"answer\":\"A\",\"type\":\"single\",\"knowledge\":[\"...\"]}]}"},
		},
	})
	if err != nil {
		return nil, err
	}
	start := strings.Index(resp.Content, "{")
	end := strings.LastIndex(resp.Content, "}")
	if start < 0 || end < 0 {
		return nil, fmt.Errorf("qbank: no json")
	}
	var raw struct {
		Questions []Question `json:"questions"`
	}
	if err := json.Unmarshal([]byte(resp.Content[start:end+1]), &raw); err != nil {
		return nil, err
	}
	for i := range raw.Questions {
		if raw.Questions[i].ID == "" {
			raw.Questions[i].ID = fmt.Sprintf("q%d", i+1)
		}
	}
	set := &Set{ID: uuid.NewString(), Title: title, Questions: raw.Questions, CreatedAt: time.Now()}
	s.mu.Lock()
	s.sets[set.ID] = set
	s.mu.Unlock()
	return set, nil
}

// Save 保存到 VFS。
func (s *Service) Save(set *Set) (string, error) {
	uri := fmt.Sprintf("vfs://qbank/%s", set.ID)
	data, _ := json.Marshal(set)
	_, err := s.vfs.Put(uri, data, map[string]string{"title": set.Title})
	return uri, err
}

// StartAttempt 开始一次练习。
func (s *Service) StartAttempt(setID string) (*Attempt, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	set, ok := s.sets[setID]
	if !ok {
		return nil, fmt.Errorf("qbank: set not found")
	}
	a := &Attempt{
		ID:        uuid.NewString(),
		SetID:     setID,
		Answers:   map[string]string{},
		Total:     len(set.Questions),
		StartedAt: time.Now(),
	}
	s.attempts[a.ID] = a
	return a, nil
}

// Answer 答题。
func (s *Service) Answer(attemptID, qID, ans string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	a, ok := s.attempts[attemptID]
	if !ok {
		return fmt.Errorf("qbank: attempt not found")
	}
	a.Answers[qID] = ans
	return nil
}

// Submit 提交并自动阅卷。
func (s *Service) Submit(attemptID string) (*Attempt, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	a, ok := s.attempts[attemptID]
	if !ok {
		return nil, fmt.Errorf("qbank: attempt not found")
	}
	set, ok := s.sets[a.SetID]
	if !ok {
		return nil, fmt.Errorf("qbank: set not found")
	}
	// 复制 questions 快照，避免后续若 set 被并发修改时触发 race
	questions := append([]Question(nil), set.Questions...)
	score := 0
	for _, q := range questions {
		if a.Answers[q.ID] == q.Answer {
			score++
			for _, k := range q.Knowledge {
				s.mastery[k] = min(100, s.mastery[k]+5)
			}
		} else {
			for _, k := range q.Knowledge {
				s.mastery[k] = max(0, s.mastery[k]-3)
			}
		}
	}
	a.Score = score
	a.FinishedAt = time.Now()
	return a, nil
}

// Analyze 深度解析（基于 AI）。
func (s *Service) Analyze(ctx context.Context, setID, qID string) (string, error) {
	s.mu.RLock()
	set, ok := s.sets[setID]
	if !ok {
		s.mu.RUnlock()
		return "", fmt.Errorf("qbank: set not found")
	}
	// 在持锁期间完成 questions 复制 + 查找目标 question
	questions := append([]Question(nil), set.Questions...)
	s.mu.RUnlock()
	var q *Question
	for i := range questions {
		if questions[i].ID == qID {
			q = &questions[i]
			break
		}
	}
	if q == nil {
		return "", fmt.Errorf("qbank: question not found")
	}
	p, _ := s.llm.Get("openai")
	if p == nil {
		return "", fmt.Errorf("qbank: no provider")
	}
	resp, err := p.Chat(ctx, llm.ChatRequest{
		Model: "gpt-4o-mini",
		Messages: []llm.Message{
			{Role: llm.RoleSystem, Content: "You are a patient tutor. Explain the problem, the knowledge point, and the solution path."},
			{Role: llm.RoleUser, Content: fmt.Sprintf("Problem: %s\nOptions: %v\nAnswer: %s", q.Stem, q.Options, q.Answer)},
		},
	})
	if err != nil {
		return "", err
	}
	return resp.Content, nil
}

// Mastery 知识点掌握度。
func (s *Service) Mastery() map[string]int {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make(map[string]int, len(s.mastery))
	for k, v := range s.mastery {
		out[k] = v
	}
	return out
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
